backend/cfg/cfg_comballoc.ml

```diff
@@ -0,0 +1,197 @@
+        { allocations = List.rev allocations; next_cell = Some cell }
+      | Reloadretaddr | Poptrap | Prologue | Pushtrap _ ->
+        (* CR-soon xclerc for xclerc: is it too conservative? *)
+        { allocations = List.rev allocations; next_cell = Some cell }
+      | Op
+          ( Move | Spill | Reload | Negf | Absf | Addf | Subf | Mulf | Divf
+          | Floatofint | Intoffloat | Valueofint | Intofvalue | Vectorcast _
+          | Opaque | Const_int _ | Const_float _ | Const_vec128 _
+          | Const_symbol _ | Stackoffset _ | Load _
+          | Store (_, _, _)
+          | Compf _ | Csel _ | Specific _ | Name_for_debugger _
+          | Probe_is_enabled _ | Scalarcast _ | Dls_get
+          | Intop
+              ( Iadd | Isub | Imul | Idiv | Imod | Iand | Ior | Ixor | Ilsl
+              | Ilsr | Iasr | Ipopcnt | Imulh _ | Iclz _ | Ictz _ | Icomp _ )
+          | Intop_imm
+              ( ( Iadd | Isub | Imul | Idiv | Imod | Iand | Ior | Ixor | Ilsl
+                | Ilsr | Iasr | Ipopcnt | Imulh _ | Iclz _ | Ictz _ | Icomp _ ),
+                _ )
+          | Intop_atomic _ ) ->
+        loop allocations (DLL.next cell) ~curr_mode ~curr_size)
+  in
+  loop [] cell ~curr_mode ~curr_size
+
+(** [combine ~max_instr_id cell] combines allocations, starting from [cell] and
+   using [max_instr_id] as the counter to get new instruction identifiers.
+
+   The allocation are combined by repeatedly:
+
+   - 1. looking for a "first" allocation;
+   - 2. looking for all subsequent allocations compatible with the "first" one;
+   - 3. continuing the process at step 1. from the instruction after the last
+        one seen at step 2.
+
+    When steps 1 and 2 are both successful, allocations are effectively
+    combined. This means that:
+
+    - the "first" allocation is made bigger to account for all allocations;
+    - the other allocations are replaced with a reference to the result of the
+      previous allocation, with a different offset. *)
+let rec combine : max_instr_id:int ref -> cell option -> unit =
+ fun ~max_instr_id cell ->
+  let first_allocation = find_next_allocation cell in
+  match first_allocation with
+  | None -> ()
+  | Some { bytes; dbginfo; mode; cell } ->
+    assert (List.length dbginfo = 1);
+    let compatible_allocs =
+      find_compatible_allocations (DLL.next cell) ~curr_mode:mode
+        ~curr_size:bytes
+    in
+    (match compatible_allocs.allocations with
+    | [] -> ()
+    | other_allocations ->
+      let first_allocation_instr = DLL.value cell in
+      let first_allocation_res0 = first_allocation_instr.res.(0) in
+      (* First, replace the "other" allocations with a reference to the result
+         of the previous allocation and compute the total size. *)
+      let total_size_of_other_allocations, dbginfo_of_other_allocations, _ =
+        List.fold_left other_allocations ~init:(0, [], first_allocation_res0)
+          ~f:(fun (size, dbginfos, prev_res0) other_allocation ->
```

Comment:
I think I prefer the current version, separating
the selection of the allocation from their
combination. As you wrote, the slight inefficiency
does not matter in practice since these list are
very small.